### DISCUSSION TITLE
Something like a pinball

### DISCUSSION THREAD

**Discussion by scenaristeur** [enable3d/enable3d#241] [Help / Questions]
Hi, bravo @yandeu for this great physic abstraction of three with ammo. 
I want to make something like a pinball but i don't understand how to do it with enable3d. 

starting from the webpack example, i've added a table, attached two cylinder and two flips. 
here is the code https://github.com/scenaristeur/flipball/blob/main/enable3D_webpack/www/scripts/three.ts
but as you can see, the left flip can rotate but does not collide with other object (the green ball) and the right flip collide with the blue cube but does not rotate when I hit the keyboard left or right arrow. 

https://user-images.githubusercontent.com/4020744/151121809-cde0214a-a1d7-4716-a0b7-b1d6ddc2ecb6.mp4

I would like to know what is the correct code . could you explain me how to implement this kind of object that can rotate and collide. 
Thxs

**Comment by yandeu** [comment]
Maybe try to make the two flips simple kinematic objects instead of working with constraints?

**Comment by scenaristeur** [comment]
thanks, Ive got something working, but always with constraints. 
could you explain "simple kinematic objects" ? 
could you just tell me what are the difference, between Locked constraint and Fixed Constraint ? 

https://user-images.githubusercontent.com/4020744/152341662-a4677635-13c2-4209-8de7-bb0132385ccc.mp4

**Reply by yandeu** [reply]
Forget about what I said. You example looks very good 👍🏻

**Comment by scenaristeur** [comment]
I want now to load real pinball parts , should i use "simple kinematic objects" too to join them together ?

![image](https://user-images.githubusercontent.com/4020744/152342380-a6cc20e8-70f6-4288-a6d6-334dfdac7826.png)

**Reply by yandeu** [reply]
Most of these object don't need physics (I presume)? So there is no need to add a physics body to it. 

I would only add physics to the parts the ball touches.

**Comment by scenaristeur** [comment]
It seems that the Yrotation axe for the "hull" shape is centered whereas the Yrotation axe for the "convex" shape is determined by the stl geometry loaded.  but the convex shape has two many vertices. Do you know how I could the hull shape, but with a good rotation axe according to the stl geometry ?

+ when the bat is rotating, the ball go  through the pinball , Could you give a reason why ? 

Thxs

https://github.com/scenaristeur/flipball/blob/main/enable3D_webpack/www/scripts/loader_test.js

![Capture d’écran du 2022-02-06 03-18-47](https://user-images.githubusercontent.com/4020744/152665526-b416eba8-6f84-460e-bf56-352cbb9ed4ac.png)
![Capture d’écran du 2022-02-06 03-18-57](https://user-images.githubusercontent.com/4020744/152665527-35ac9ad4-c478-4dc6-8bb2-bfbd0923ea56.png)

**Comment by scenaristeur** [comment]
@yandeu 

I am now with this code https://github.com/scenaristeur/flipball/blob/main/enable3D_webpack/www/scripts/loader_throw.js

With just the stl rubber that flip with arrows and down arrow to have new balls. But sometimes the ball pass through the rubber . I 've tried with the ccd function but do i have to apply to the two ( the ball and the rubber) because this is the rubber that have a high velocity ? What parameter should i use to the rubber as the shape is not a standard (sphere or box) ?

**Comment by scenaristeur** [comment]
reply to https://github.com/enable3d/enable3d/discussions/337#discussioncomment-4066373
> The Marble and Modele do not seem to work. The Flipball works great but is cut off at the bottom. Might want to fix it first?

app here https://scenaristeur.github.io/flipball/
code here https://github.com/scenaristeur/flipball
 > The Flipball works great but is cut off at the bottom

--> you can  zoom out with two finger and rotate your table as you want
> The Marble and Modele do not seem to work


--> marble is an attemp to build a marble machine like this one https://www.youtube.com/watch?v=0b-B_Bg31Zg. 
each time a ball collide a box, a sound is emmitted
but many sound to load take a long time to start correctly, perharps remove some boxes 

--> modele is an empty-sample file to show how to build a custom table in https://github.com/scenaristeur/flipball/tree/main/src/tables
